### PR TITLE
[RPC] Allow to explicitly define simple function struct

### PR DESCRIPTION
**Pull request by mbasmanova**

### CODE REVIEW

File: velox/functions/prestosql/DateTimeFunctions.h [hunk 1]
```diff
@@ -118,23 +118,25 @@ FOLLY_ALWAYS_INLINE bool call(
 }
 VELOX_UDF_END();
 
-VELOX_UDF_BEGIN(hour)
-const date::time_zone* timeZone_ = nullptr;
+struct udf_hour {
```
Comment (damianr99) [inline velox/functions/prestosql/DateTimeFunctions.h:121]: I like that you can see the outer struct wrapper, and you can see where the "udf_hour" symbol comes from.

File: velox/functions/prestosql/DateTimeFunctions.h [hunk 2]
```diff
@@ -118,23 +118,25 @@ FOLLY_ALWAYS_INLINE bool call(
...
+  VELOX_UDF_BEGIN2(hour)
```
Comment (damianr99) [inline velox/functions/prestosql/DateTimeFunctions.h:122]: it's a shame you have to repeat the "hour" here.  I imagine confusing things happen if you get this out of sync with the name above?  Is there some way of removing that.  Looking at VELOX_UDF_BEGIN2 it only seems to use it in one place to define a string for `static constexpr auto name`.  

Another variant that might spark an even better idea, you could get make the struct visible and not repeat the name using something like:

struct VELOX_UDF("hour") {
   ...
}

you don't even need the VELOX_UDF_END as it's just a } and it's obvious you need it and what scope it encloses.

Comment (damianr99) [reply velox/functions/prestosql/DateTimeFunctions.h:122]: actually ignore that, I just realized that won't work, the user supplied code will be in an extra level of nesting compared to what VELOX_UDF expands to. :(

File: velox/functions/Macros.h [hunk 3]
```diff
@@ -66,6 +66,51 @@
   }                     \
   ;
 
+#define VELOX_UDF_BEGIN2(Name)                                                 \
+  template <typename __Velox_ExecParams>                                       \
+  struct udf {                                                                 \
```
Comment (funrollloops) [inline velox/functions/Macros.h:71]: Ah, I had forgotten that the macros actually define a struct _within_ a struct. It'll be tricky/fragile to use inheritance to share functionality between UDFs in this framework, I agree.

Comment (mbasmanova) [reply velox/functions/Macros.h:71]: I wanted to mention this to you, but you beat me to it :-)

Comment (mbasmanova) [reply velox/functions/Macros.h:71]: @funrollloops Sagar, do you see value in changing the macros like this to make it more explicit in the code that a function is a struct?

Comment (funrollloops) [reply velox/functions/Macros.h:71]: Making the type of the top-level thing visible is somewhat useful, but the benefit is pretty marginal IMO. On the flip side people may make the mistake I made of thinking they could inherit + override, when that won't actually work.

**Pull request closed without merging**